### DISCUSSION TITLE
Add support for creating hodl invoices

### DISCUSSION THREAD

**Discussion by aaronkopplin** [btcpayserver/btcpayserver#5058] [Ideas / Feature Requests]
Hi there,

I would like it if BTC pay server could create hold invoices. These invoices would allow for free refunds without multiple lightning transactions, and would save on fees. Without hold invoices, stores are vulnerable to a type of attack where the attacker creates an order, sends a payment, then requests a refund. This requires two payments and the attacker could have a routing node in the middle, collecting routing fees from the merchant. 

for those that arent familiar with hold invoices, here is a good primer:

"Hold invoices are LN invoices where the receiver doesn’t immediately release the preimage upon receiving a payment. Instead, the receiver performs some action and then either accepts the payment, explicitly rejects it, or lets it time out.

For example, Alice could automatically generate hold invoices on her website but wait until a customer actually paid before searching her inventory for the requested item. This would give her a chance to cancel the payment if she couldn’t deliver."

source: https://bitcoinops.org/en/topics/hold-invoices/#:~:text=Hold%20invoices%20are%20LN%20invoices,or%20lets%20it%20time%20out.

A longer explanation: https://thebitcoinmanual.com/articles/hodl-invoice/

Hold invoices also allow for other beneficial types of transactions. If anyone knows of any benefits, please post them here. Thanks!

**Comment by Johnathanodom** [comment]
So, I must wait for a bit?

**Comment by dennisreimann** [comment]
Right now we don't support hold invoices in BTCPay yet, but it would be doable – though it would require LND.

**Comment by mp21-org** [comment]
I like the feature as well, but would hope that it's implementation in bolt12 would be more resiliant.
Maybe we can keep this open until we have bolt12?

**Reply by Kukks** [reply]
Bolt12 doesn't do hold invoices but instead creates a secondary refund invoice, similar to lnurlw

**Comment by Kukks** [comment]
> Hi there,
> 
> I would like it if BTC pay server could create hold invoices. These invoices would allow for free refunds without multiple lightning transactions, and would save on fees. Without hold invoices, stores are vulnerable to a type of attack where the attacker creates an order, sends a payment, then requests a refund. This requires two payments and the attacker could have a routing node in the middle, collecting routing fees from the merchant.
> 
> for those that arent familiar with hold invoices, here is a good primer:
> 
> "Hold invoices are LN invoices where the receiver doesn’t immediately release the preimage upon receiving a payment. Instead, the receiver performs some action and then either accepts the payment, explicitly rejects it, or lets it time out.
> 
> For example, Alice could automatically generate hold invoices on her website but wait until a customer actually paid before searching her inventory for the requested item. This would give her a chance to cancel the payment if she couldn’t deliver."
> 
> source: [bitcoinops.org/en/topics/hold-invoices/#:~:text=Hold%20invoices%20are%20LN%20invoices,or%20lets%20it%20time%20out](https://bitcoinops.org/en/topics/hold-invoices/#:~:text=Hold%20invoices%20are%20LN%20invoices,or%20lets%20it%20time%20out).
> 
> A longer explanation: [thebitcoinmanual.com/articles/hodl-invoice](https://thebitcoinmanual.com/articles/hodl-invoice/)
> 
> Hold invoices also allow for other beneficial types of transactions. If anyone knows of any benefits, please post them here. Thanks!

Hold invoices do not scale and should be discouraged for use if anything. For every hold invoice that is pending releasing the preimage, there is liquidity locked up along the entire route, and there is a limited amount of in-flight HTLCs that each channel can use, around ~500, and when reached, makes the channel unusable.

Requesting a refund is not an attack, you should charge a refund processing fee to discourage that behavior if it becomes a problem.

**Reply by aaronkopplin** [reply]
Hi Kukks. I understand your point. Would this fact be mitigated by the fact that there are relatively few _stores_ in comparison to lightning users? I find it hard to believe that regular lightning users would be making a lot of hold invoices. Additionally, not every store would be accepting returns via hold invoices. Therefore, the actual market segment (in terms of all of lightning) is a fraction of a fraction of total users. So, liquidity issues would not be as dramatic as if every use would be using hold invoices frequently. Does this impact your thinking at all? Thanks for the consideration.
